---
title: "Untitled"
output: html_document
date: "2025-08-07"
---

## Configure inputs
```{r, Configure several inputs}
timestampFiles <- format(Sys.time(), "%Y%m%d.%H%M")

# GSEA software path
pathToBatIn <- "C:/Program Files/GSEA_4.2.3"

# Folder with files downloaded from GSEA, mSigDB, or other external
# sources (.chip and .gmt files)
chipDir <- "C:/GSEA_R_Demo/chip"
chipFile <- file.path(chipDir, "Human_NCBI_Gene_ID_MSigDB.v2023.2.Hs.chip")  

gmtDir <- "C:/GSEA_R_Demo/gmt/msigdb_v2023.2.Hs_GMTs"

# .gmt Gene identifier
gmtFileType <- ".v2023.2.Hs.symbols.gmt"

# list with prefixes of desired gene set collections to perform GSEA across comparisons
gmtsKeep <- c("h.all", "c2.cp")

# Dataset parameters
setName <- "GSE181194_UM_BAP1_Mut_WT"
outDir <- file.path("C:/GSEA_R_Demo", setName)

# GSEA results folder
gseaResDir <- file.path(outDir, "GSEA", "results")
dir.create(gseaResDir, recursive = T)

# GSEA data folder
gseaDataDir <- file.path(outDir, "GSEA", "data")
dir.create(gseaDataDir)

# GSEA-ready file names
fileNameOutGCT <- file.path(gseaDataDir, paste(timestampFiles, setName, "gct", sep = "."))
fileNameOutcls <- file.path(gseaDataDir, paste(timestampFiles, setName, "cls", sep = "."))

```

## Code from GEO2R
```{r, Example GEO2R implementation}
# Version info: R 4.2.2, Biobase 2.58.0, GEOquery 2.66.0, limma 3.54.0
################################################################
#   Differential expression analysis with DESeq2
library(DESeq2)

# load counts table from GEO
urld <- "https://www.ncbi.nlm.nih.gov/geo/download/?format=file&type=rnaseq_counts"
path <- paste(urld, "acc=GSE181194", "file=GSE181194_raw_counts_GRCh38.p13_NCBI.tsv.gz", sep="&");
tbl <- as.matrix(data.table::fread(path, header=T, colClasses="integer"), rownames="GeneID")

# load gene annotations 
apath <- paste(urld, "type=rnaseq_counts", "file=Human.GRCh38.p13.annot.tsv.gz", sep="&")
annot <- data.table::fread(apath, header=T, quote="", stringsAsFactors=F, data.table=F)
rownames(annot) <- annot$GeneID

# sample selection
gsms <- "000111222"
sml <- strsplit(gsms, split="")[[1]]

# group membership for samples
gs <- factor(sml)
groups <- make.names(c("WT_92_1","MUT_WM1366","MUT_MM28"))
levels(gs) <- groups
sample_info <- data.frame(Group = gs, row.names = colnames(tbl))

# pre-filter low count genes
# keep genes with at least N counts > 10, where N = size of smallest group
keep <- rowSums( tbl >= 10 ) >= min(table(gs))
tbl <- tbl[keep, ]

ds <- DESeqDataSetFromMatrix(countData=tbl, colData=sample_info, design= ~Group)

ds <- DESeq(ds, test="LRT", reduced = ~ 1)  # Use LRT for all-around gene ranking

# extract results for top genes table
r <- results (ds, alpha=0.05, pAdjustMethod ="fdr")

tT <- r[order(r$padj)[1:250],] 
tT <- merge(as.data.frame(tT), annot, by=0, sort=F)

tT <- subset(tT, select=c("GeneID","padj","pvalue","stat","baseMean","Symbol","Description"))
write.table(tT, file=stdout(), row.names=F, sep="\t")

plotDispEsts(ds, main="GSE181194 Dispersion Estimates")

# create histogram plot of p-values
hist(r$padj, breaks=seq(0, 1, length = 21), col = "grey", border = "white", 
         xlab = "", ylab = "", main = "GSE181194 Frequencies of padj-values")

# Wald test to obtain contrast-specific results
ds <- DESeq(ds, test="Wald", sfType="poscount")
r <- results (ds, contrast=c("Group", groups[1], groups[2]), alpha=0.05, pAdjustMethod = "fdr")

# volcano plot
old.pal <- palette(c("#00BFFF", "#FF3030")) # low-hi colors
par(mar=c(4,4,2,1), cex.main=1.5)
plot(r$log2FoldChange, -log10(r$padj), main=paste(groups[1], "vs", groups[2]),
     xlab="log2FC", ylab="-log10(Padj)", pch=20, cex=0.5)
with(subset(r, padj<0.05 & abs(log2FoldChange) >= 0),
     points(log2FoldChange, -log10(padj), pch=20, col=(sign(log2FoldChange) + 3)/2, cex=1))
legend("bottomleft", title=paste("Padj<", 0.05, sep=""), legend=c("down", "up"), pch=20,col=1:2)

# MD plot
par(mar=c(4,4,2,1), cex.main=1.5)
plot(log10(r$baseMean), r$log2FoldChange, main=paste(groups[1], "vs", groups[2]),
     xlab="log10(mean of normalized counts)", ylab="log2FoldChange", pch=20, cex=0.5)
with(subset(r, padj<0.05 & abs(log2FoldChange) >= 0),
     points(log10(baseMean), log2FoldChange, pch=20, col=(sign(log2FoldChange) + 3)/2, cex=1))
legend("bottomleft", title=paste("Padj<", 0.05, sep=""), legend=c("down", "up"), pch=20,col=1:2)
abline(h=0)
palette(old.pal) # restore palette

# Venn diagram
library(gplots)
all_res <- list()

ct.names <- resultsNames(ds)[-1] # contrasts names without Intercept
for (ct in ct.names) {
  r <- results(ds, name=ct, alpha=0.05, pAdjustMethod = "fdr")
  all_res[[length(all_res) + 1]] <- rownames(r)[!is.na(r$padj) & r$padj < 0.05 & abs(r$log2FoldChange) >= 0]
}
names(all_res) <- ct.names

venn(all_res)

################################################################
#   General expression data visualization
dat <- log10(counts(ds, normalized = T) + 1) # extract normalized counts

# box-and-whisker plot
lbl <- "log10(raw counts + 1)"
ord <- order(gs)  # order samples by group
palette(c("#1B9E77", "#7570B3", "#E7298A", "#E6AB02", "#D95F02",
          "#66A61E", "#A6761D", "#B32424", "#B324B3", "#666666"))
par(mar=c(7,4,2,1))
boxplot(dat[,ord], boxwex=0.6, notch=T, main="GSE181194", ylab="lg(norm.counts)", outline=F, las=2, col=gs[ord])
legend("topleft", groups, fill=palette(), bty="n")

# UMAP plot (multi-dimensional scaling)
library(umap)
dat <- dat[!duplicated(dat), ] # first remove duplicates
par(mar=c(3,3,2,6), xpd=TRUE, cex.main=1.5)
ump <- umap(t(dat), n_neighbors = 4, random_state = 123)
plot(ump$layout, main="UMAP plot, nbrs=4", xlab="", ylab="", col=gs, pch=20, cex=1.5)
legend("topright", inset=c(-0.15,0), legend=groups, pch=20,
       col=1:length(groups), title="Group", pt.cex=1.5)
```


## GSEA prep
```{r, Build comparisons - adjust as needed}
# pAnnot = sample_info groups = levels(pAnnot$Group)
groups <- levels(sample_info$Group)
allCombs <- combn(groups, 2)

colNameContrast <- "Group"

# Helper to make 'control' the denominator (row)
allCombs[, ] <- allCombs[c(2, 1), ]

# Comparison names
tempCompColAdd <- paste0(allCombs[1, ], ".vs.", allCombs[2, ])

# build comparison df
contrasts_df <- data.frame(column = NA, denom = allCombs[2, ], 
                           numer = allCombs[1, ], compLbl = tempCompColAdd)
contrasts_df$column <- colNameContrast

# Prealloc gct and cls file names
contrasts_df$gct <- NA
contrasts_df$cls <- NA

# Nomenclature for specific comparison via GSEA
contrasts_df$clsComparison <- paste0(contrasts_df$numer, "_versus_", contrasts_df$denom)

# Set chip file
contrasts_df$chipFile <- chipFile

```

## Generate cls and gct files
```{r, Generate cls and gct files}
# gct-ready dds data object
dds <- DESeqDataSetFromMatrix(countData = tbl, colData = sample_info, design = ~Group)

#' tbl is reduced within GEO2R code to genes with rowSums(counts) > 10, but keeping this code 
#' here so it can be adapted for other pipelines easily 
# countsMin = 1
# keep <- rowSums(counts(dds)) >= countsMin
# dds_gct <- DESeq(dds[keep,])
# rm(keep)

dds_gct <- DESeq(dds)

# Create gct object from DESeq2 normalized counts data
normcts_gct <- counts(dds_gct, normalized = TRUE)
normctsGCT <- new("GCT", mat = normcts_gct, rdesc = as.data.frame(rowData(dds_gct)),
    cdesc = as.data.frame(colData(dds_gct)))
rm(dds_gct, normcts_gct)

# Write gct data to file
write_gct(normctsGCT, fileNameOutGCT, ver = 2, appenddim = F)

# add filename to comparisons df
contrasts_df$gct <- fileNameOutGCT

# Write cls data to file
write_cls(normctsGCT@cdesc[, colNameContrast, drop = F], groupingColName = colNameContrast,
    filenameOut = fileNameOutcls)

# add filename to comparisons df
contrasts_df$cls <- fileNameOutcls

```

## Subset gmt files to include
```{r, data frame with gmt files and some GSEA parameters - create df with abbreviations}
# Get list of gmt files
gmtFileIn <- dir(gmtDir, pattern = gmtFileType, full.names = TRUE, recursive = F)
names(gmtFileIn) <- gsub(gmtFileType, "", basename(gmtFileIn))
gmtDF <- as.data.frame(gmtFileIn)

gmtDF <- gmtDF[gmtsKeep, , drop = F]

# Add some parameters for specific gene sets
gmtDF$plot_top_x <- 50
gmtDF$maxGenesInSet <- 500

# Customize parameters for gene set collection(s)
gmtDF[rownames(gmtDF) %in% "h.all", "plot_top_x"] <- 100  # to plot all gene sets

```

## Run GSEA
```{r, Run GSEA via command prompt for each comparison and each gmt file}

# For each contrast, run comparisons for gmts
for (j in 1:nrow(contrasts_df)) {

    # Print comparison and time
    paste0(contrasts_df$compLbl[j], " - ", format(Sys.time(), "%H:%M"))

    # Set output directory
    GSEAresOutDir <- file.path(gseaResDir, contrasts_df$compLbl[j])

    # Create directory if it does not exist
    if (!dir.exists(GSEAresOutDir)) {
        dir.create(GSEAresOutDir, recursive = TRUE)
    }

    # Run for each gmt file
    for (i in 1:nrow(gmtDF)) {

        # Get java usage
        temp_javaVal <- checkjava()
        
        if (i == 1 & j == 1){
          if (length(temp_javaVal) > 0){
            stop("java is already running. This script will not run GSEA if java is already running.")
          }
        }

        # If java is running, pause for set amount of time to prevent from every analysis being submitted at the same ~time
        while (length(temp_javaVal) > 0) {
            print(paste0(format(Sys.time(), "%H:%M:%S"), " - waiting 30s"))
            Sys.sleep(time = 30)
            temp_javaVal <- checkjava()
        }

        # Temp abbreviation
        GSEArepLbl_tmp <- paste0(rownames(gmtDF)[i], "_", contrasts_df$compLbl[j])
        print(paste0(GSEArepLbl_tmp, " - ", format(Sys.time(), "%H:%M")))

        # Run GSEA 
        runGSEA(pathToBat = pathToBatIn, gctExprfile = contrasts_df$gct[j],
            clsLabelsFile = contrasts_df$cls[j], clsComparison = contrasts_df$clsComparison[j],
            GSEArepLbl = GSEArepLbl_tmp, GSEAresOutDir = GSEAresOutDir,
            gmtFile = gmtDF$gmtFileIn[], set_max = gmtDF$maxGenesInSet[],
            zip_report = F, ChipFile = contrasts_df$chipFile[j], plot_top_x = gmtDF$plot_top_x[i])
    }
}
  

```

## Save contrasts for re-running GSEA on other gene set collections
```{r}
fnOutContrasts <- file.path(outDir, paste(timestampFiles, setName, "contrasts", "tsv", sep = "."))
write.table(contrasts_df, fnOutContrasts, sep = "\t", row.names = F, col.names = T)

```


